Migrate password hashing away from mysql password()

tomtom5152 opened the pull request:
This is deprecated and needs to be replaced. 
Use [passlib](http://passlib.readthedocs.io/en/stable/index.html) to hash, store, verify and upgrade as necessary.
Includes patch to migrate existing passwords in a non-breaking way.
__Auth.salt is also redundant and removed as passlib stores the salt in the complete hash string.
Tests have been minimally updated as needed to reflect breaking internal changes, no modifications were made to the existing test framework.

Fixes #5195

tomtom5152 committed:
Migrate password hashing away from mysql password()

This is deprecated and needs to be replaced. 
Use passlib to hash, store, verify and upgrade as necessary.
Includes patch to migrate existing passwords in a non-breaking way.

Fixes #5195

tomtom5152 commented:
Any chance of some movement on this please? Have rebased onto develop which hopefully will have fixed the issues with unrelated tests failing and resolved the merge conflict.

tomtom5152 committed:
Fix code quality issue

Trailing whitespace in patch SQL, was unoticable in editor.

codingCoffee commented:
Give me a day more at the most. I'll look into it.

codingCoffee reviewed (changes requested):
Review:

Tested with multiple passwords containing A-Z, a-z, 0-9, /*-... and Extended ASCII, migration (patch) works perfectly.

The __Auth table before and after (for maintainers):

![screenshot_20180411_095108](https://user-images.githubusercontent.com/13611153/38596216-dee618e2-3d6d-11e8-846d-108c3f5dc48d.png)

To sum it up a great PR, and a much needed one! :clap: 
Those minor changes should make it mergable.

tomtom5152 committed:
Fix PR issues from #5200.

Fix issue with with rebase messing up patches.txt.

Change auth parameter back to user.

tomtom5152 commented:
Changes made as request @codingCoffee. There was a deliberate reasoning behind the change from user to auth, but I can't remember for the life of me what that is now. I've also been programming in Go recently where parameter name makes no difference, but I suppose it does in Python because kwargs are a thing so changed it back.

It also appears I properly messed up the rebase hence patches.txt being such a mess, that's been fixed.

codingCoffee commented:
@tomtom5152 awesome :+1:

rmehta committed:
Update migrate_passwords_passlib.py

"code is poetry"

rmehta committed:
Update password.py

rmehta committed:
Merge branch 'develop' into password-hashing

codingCoffee committed:
Merge branch 'develop' into password-hashing

codingCoffee commented:
@rmehta merga-able :+1:
